### PR TITLE
Dev worksheet 4

### PR DESCRIPTION
**Pull request by LivanKov**
Pull Request for Assignment 4
**Summary:**
- Implemented all missing features from the past assignment
- Implemented a thermostat along with thorough unit tests
- Enabled running Rayleigh-Taylor Instability
- Implemented input for gravity, epsilon, delta
- Implemented Periodic Boundary Conditions via the "Ghost Particle" method
- Implemented a class for "checkpointing"
- Ran the "falling drop" simulations
- Used GNU prof for performance assessment
- Main optimizations currently come from naive algorithm optimization and trying to prioritize in-place object construction
- Refactoring for a more structured code

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -35,27 +35,148 @@ $ make
...
+```
 
+In this example:
+- **-i ../input_file.xml** is the XML file that provides the base configuration for the simulation.
+- **-e 100.0** overrides the `end_time` specified in the XML file to run the simulation for 100.0 units of time.
+- **-d 0.01** overrides the `delta_time` specified in the XML file to set the time increment to 0.01.
+- **-o ../output/simulation_output** overrides the `output_path` specified in the XML file to set the output file path.
+- **-t 5** overrides the `write_frequence` specified in the XML file to set the write frequency to write a file every 5 iterations.
+
+### EML Example
```
Comment (manishmishra6016) [inline README.md:83]:
```suggestion
### XML Example
```

File: src/io/output/CheckpointWriter.h [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  ~CheckpointWriter();
+
+  /**
+   * @brief Writes a checkpoint file from the given particle container.
+   * @param particles A reference to the particle container to write from.
+   * @param filename The name of the file to write to.
+   * @param delta_t The delta_t value to write.
+   * @param t_end The t_end value to write.
+   */
+  static void writeCheckpoint(LinkedCellContainer &particles,
```
Comment (manishmishra6016) [inline src/io/output/CheckpointWriter.h:26]: So, it seems like checkpointing will not work with `DirectSum`. Also, you removed the `ParticleContainer` interface, so both DirectSum and LinkedCells are independent now, otherwise, you could have used something `shared_ptr` here to make this work with both.

File: src/simulator/Thermostat.cpp [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+  // in the very end we apply our scaling factor to the particles
+  /* for (auto& p : particles_) {
+       auto& current_velocity = p.getV();
+
+       std::array<double, 3> new_velocity{};
+       for (size_t i = 0; i < dimensions_; ++i) {
+           new_velocity[i] = current_velocity[i] * scaling_factor_;
+       }
+       p.updateV(new_velocity);
+   } */
```
Comment (manishmishra6016) [inline src/simulator/Thermostat.cpp:102]: dead code

File: src/simulator/Thermostat.cpp [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+      std::array<double, 3> random_velocity =
+  maxwellBoltzmannDistributedVelocity(average_velocity, dimensions_);
+
+      auto current_velocity = particle.getV();
+
+      for (size_t i = 0; i < dimensions_; ++i) {
+          current_velocity[i] += random_velocity[i];
+      }
+      particle.updateV(current_velocity);
+  } */
```
Comment (manishmishra6016) [inline src/simulator/Thermostat.cpp:183]: dead code

File: src/simulator/Thermostat.cpp [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+  // in the very end we apply our scaling factor to the particles
+  /*for (auto& p : particles_) {
+      auto& current_velocity = p.getV();
+
+      std::array<double, 3> new_velocity{};
+      for (size_t i = 0; i < 3; ++i) {
+          new_velocity[i] = current_velocity[i] * scaling_factor_;
+      }
+      p.updateV(new_velocity);
+  } */
```
Comment (manishmishra6016) [inline src/simulator/Thermostat.cpp:234]: dead code, just FYI, if you want to keep another version of the code, the best practice is to simply keep it on another branch. And if you no longer use, just remove it, you can always access it with the older commit IDs.

File: src/simulator/Thermostat.cpp [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+    p.updateV(new_velocity);
+  }
+}
+
+// ------------- getters  --------------------------------------------
+
+double Thermostat::get_current_temperature() {
+  calculate_current_temperature();
+  return current_temperature_;
+}
```
Comment (manishmishra6016) [inline src/simulator/Thermostat.cpp:252]: Here, the function `calculate_current_temperature()` is not really needed, you can always just use the `get_current_temperature()`. This might reduce unnecessary function call. 

Another (nicer, IMO) way would be to not do the calculate in this getter function and instead just return. Basically, separate computation and just access. This would also help a bit as then this function can be inlined. I am not sure about how much advantage this would give but it just makes more sense to me. Having said, it is not too critical.

File: src/simulator/Simulation.cpp [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+  }
+
+  // Start measuring time for the main simulation loop
+  auto start_time = std::chrono::high_resolution_clock::now();
+
+  while (current_time < params_.end_time) {
+
+    size_t molecules_this_iteration = particles.size();
+
+    Calculation<Position>::run(particles, params_.time_delta, option);
```
Comment (manishmishra6016) [inline src/simulator/Simulation.cpp:96]: When you read from the checkpoint, there is no `force` available from the previous step. So, you need to compute it once before starting the main loop. Currently, the very first step (position update) after reading the checkpoint file is missing the force contribution.

File: src/simulator/calculations/Velocity.cpp [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+    if (SimParams::enable_v_threshold) {
+      // Clamp velocity based on absolute threshold (positive and negative)
+      for (int i = 0; i < 3; ++i) {
+        if (v[i] > SimParams::v_threshold) {
+          v[i] = SimParams::v_threshold; // Positive threshold
+        } else if (v[i] < -SimParams::v_threshold) {
+          v[i] = -SimParams::v_threshold; // Negative threshold
+        }
+      }
+    }
```
Comment (manishmishra6016) [inline src/simulator/calculations/Velocity.cpp:19]: Why is this required? 
This would destroy the symplectic nature of the time integration scheme we use, so I am wondering what was the use case here.

File: src/simulator/calculations/Force.cpp [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+
+      if (distance > 1e-5) {
+
+        double totalForce;
+        double sigma;
+        if (it->first->getSigma() == it->second->getSigma()) {
+          sigma = it->first->getSigma();
+        } else {
+          sigma = (it->first->getSigma() + it->second->getSigma()) / 2;
+        }
```
Comment (manishmishra6016) [inline src/simulator/calculations/Force.cpp:41]: you could try some optimization here if time permits as discussed in the last review, like precomputing possible combinations and using them directly instead of computing for all pairs separately.

File: tests/periodic_conditions_test.cpp [hunk 10]
```diff
@@ -0,0 +1,571 @@
...
+    for (const auto& n : upper_right_neighbours) {
+        upper_right_actual_ids.push_back(n->getType());
+    }
+    std::sort(upper_right_actual_ids.begin(), upper_right_actual_ids.end());
+    std::sort(upper_right_expected_ids.begin(), upper_right_expected_ids.end());
+    
+    EXPECT_EQ(upper_right_actual_ids, upper_right_expected_ids) << "Upper right neighbor IDs do not match expected values";
+
+}
+*/
```
Comment (manishmishra6016) [inline tests/periodic_conditions_test.cpp:168]: dead code - why did you comment these tests?

File: tests/periodic_conditions_test.cpp [hunk 11]
```diff
@@ -0,0 +1,571 @@
...
+
+TEST_F(PeriodicBoundaryTest, PeriodicTransitionTest) {
```
Comment (manishmishra6016) [inline tests/periodic_conditions_test.cpp:170]: No comment provided about what this test is really doing. Briefly define the goal of the test and expected output.

**Review by manishmishra6016 — COMMENTED** [review]
**Note**
_This is only feedback regarding the code directly. Grade feedback will be on Moodle once we are through with all groups._

-----------------------------------------------------------------------------------------------------

Thank you for your submission :muscle: 
I think you were able to meet most of the objectives in terms of the implementation. I would have liked to see more detailed profiling results and effect of optimizations you carried out. Also, please avoid a lot of dead codes, and provide comments on the tests. 

I am currently trying to build the code on linux cluster and it doesn't seem to work right away. Can you confirm that you were able to build this exact version of the code or did you need to make some adjustments (to cmake file for instance)? I haven't looked at it closely yet, but if you have some comments that can be helpful to me, I would appreciate it.

Have a look at my comments and see if they make sense. Feel free to reply/ clarify something if needed. I wish you good luck with the final worksheet. :smile:

Cheers,
Manish

**Comment by manishmishra6016** [comment]
Just FYI,
I was able to compile on the cluster with some minor changes to cmake file.

I added ``` find_package(XercesC REQUIRED)```
and changed
``` target_link_libraries(MolSim PUBLIC xerces-c)``` to ```target_link_libraries(MolSim PUBLIC XercesC::XercesC)```

With these, it works seamlessly with `cmake ..` followed by `make` in the build folder.

I was not able to compile with the intel compiler however. This is I think due to the use of C++20 features "concepts" (+ maybe some other as well).

**Comment by LivanKov** [comment]
> Just FYI, I was able to compile on the cluster with some minor changes to cmake file.
> 
> I added ` find_package(XercesC REQUIRED)` and changed ` target_link_libraries(MolSim PUBLIC xerces-c)` to `target_link_libraries(MolSim PUBLIC XercesC::XercesC)`
> 
> With these, it works seamlessly with `cmake ..` followed by `make` in the build folder.
> 
> I was not able to compile with the intel compiler however. This is I think due to the use of C++20 features "concepts" (+ maybe some other as well).

Interesting, I wasn't the one doing the benchmarks on the cluster, so I'd have to further clarify that with my teammates. From what I've gathered however, they had no issues with compiling and running the program. 
As far as the concepts go, they aren't strictly necessary but they do enforce slightly stricter behaviour for templated classes, which I like, so I just decided to leave them in. 
Also, if I recall correctly, during our last meeting, it has been clarified that the updated cluster that we were supposed to use, would generally have no problems compiling code with more modern features in it.

If you have any more questions/complaints, please let me know.

Cheers

File: src/io/output/CheckpointWriter.h (hunk 2 above)
Comment (LivanKov) [reply src/io/output/CheckpointWriter.h:26]: LinkedCellContainer serves as a wrapper for the DirectSumContainer. 
It can still have the exact same functionality by omitting the "cells" array.
